File: src/chapters/minimization/content/code/simplex-basic.go
package main

import (
	"fmt"
	"math"
)

func dist(a, b []float64) float64 {
	s := 0.0
	for i := range a {
		s += (a[i] - b[i]) * (a[i] - b[i])
	}
	return math.Sqrt(s)
}

// Basic fixed-shape simplex method (reflect worst through centroid, else shrink to best).
func simplexBasic(f func([]float64) float64, x0 []float64, step, tol float64, maxIter int) []float64 {
	n := len(x0)
	P := make([][]float64, n+1)
	for i := range P {
		P[i] = append([]float64{}, x0...)
		if i > 0 {
			P[i][i-1] += step
		}
	}
	fv := make([]float64, n+1)
	for i := range P {
		fv[i] = f(P[i])
	}
	for it := 0; it < maxIter; it++ {
		iw, ib := 0, 0
		for i := 1; i <= n; i++ {
			if fv[i] > fv[iw] {
				iw = i
			}
			if fv[i] < fv[ib] {
				ib = i
			}
		}
		sz := 0.0
		for i := 0; i <= n; i++ {
			sz = math.Max(sz, dist(P[i], P[ib]))
		}
		if sz < tol {
			break
		}
		c := make([]float64, n) // centroid of all but the worst
		for i := 0; i <= n; i++ {
			if i != iw {
				for j := 0; j < n; j++ {
					c[j] += P[i][j] / float64(n)
				}
			}
		}
		xr := make([]float64, n) // reflect the worst vertex
		for j := 0; j < n; j++ {
			xr[j] = c[j] + (c[j] - P[iw][j])
		}
		fr := f(xr)
		if fr < fv[iw] {
			P[iw] = xr
			fv[iw] = fr
		} else { // shrink toward the best
			best := append([]float64{}, P[ib]...)
			for i := 0; i <= n; i++ {
				if i != ib {
					for j := 0; j < n; j++ {
						P[i][j] = best[j] + 0.5*(P[i][j]-best[j])
					}
					fv[i] = f(P[i])
				}
			}
		}
	}
	ib := 0
	for i := 1; i <= n; i++ {
		if fv[i] < fv[ib] {
			ib = i
		}
	}
	return P[ib]
}

func main() {
	f := func(v []float64) float64 { return (v[0]-1)*(v[0]-1) + (v[1]-2)*(v[1]-2) }
	fmt.Println(simplexBasic(f, []float64{0, 0}, 1, 1e-8, 500)) // -> [1 2]
}

File: src/chapters/minimization/content/code/simplex-basic.js
const dist = (a, b) => Math.sqrt(a.reduce((s, x, i) => s + (x - b[i]) ** 2, 0));

// Basic fixed-shape simplex method (reflect worst through centroid, else shrink to best).
function simplexBasic(f, x0, step = 1, tol = 1e-8, maxIter = 500) {
  const n = x0.length;
  const P = [x0.slice(), ...Array.from({ length: n }, (_, i) => { const v = x0.slice(); v[i] += step; return v; })];
  const fv = P.map(f);
  for (let it = 0; it < maxIter; it++) {
    let iw = 0, ib = 0;
    for (let i = 1; i <= n; i++) { if (fv[i] > fv[iw]) iw = i; if (fv[i] < fv[ib]) ib = i; }
    let sz = 0;
    for (let i = 0; i <= n; i++) sz = Math.max(sz, dist(P[i], P[ib]));
    if (sz < tol) break;
    const c = Array(n).fill(0);                                  // centroid of all but the worst
    for (let i = 0; i <= n; i++) if (i !== iw) for (let j = 0; j < n; j++) c[j] += P[i][j] / n;
    const xr = c.map((cj, j) => cj + (cj - P[iw][j]));           // reflect the worst vertex
    const fr = f(xr);
    if (fr < fv[iw]) { P[iw] = xr; fv[iw] = fr; }
    else {                                                       // shrink toward the best
      const best = P[ib].slice();
      for (let i = 0; i <= n; i++) if (i !== ib) {
        P[i] = P[i].map((pij, j) => best[j] + 0.5 * (pij - best[j]));
        fv[i] = f(P[i]);
      }
    }
  }
  let ib = 0;
  for (let i = 1; i <= n; i++) if (fv[i] < fv[ib]) ib = i;
  return P[ib];
}

const f = (v) => (v[0] - 1) ** 2 + (v[1] - 2) ** 2;
console.log(simplexBasic(f, [0, 0]));                            // -> [1, 2]
